### PR TITLE
Tache5.1_ThomasL

### PR DESCRIPTION
**Pull request by thomasleydet**
All the algorithms below will allow the user to play with the hyperparameters to find a better output result.

agnes.py: linkage_method is the linkage method used to calculate the distance between clusters, n_clusters is the number of clusters to form, and dist_metric is the distance metric used between points.

gmm.py: the function takes a dataframe and a range of number of components for the Gaussian mixture model. It verifies that the data is numeric and without missing values, and returns the fitted GMM model with the optimal number of components selected by the BIC. Data is normalized with StandardScaler before model fitting. The function also plots the BIC as a function of the number of components and chooses the optimal number of components by minimizing the BIC.

dbscan.py: This function takes as input a pandas DataFrame or numpy array and returns the cluster labels assigned by DBSCAN along with the corresponding silhouette score. The eps and min_samples parameters are DBSCAN's hyperparameters that define the maximum distance between samples for them to be considered part of the same neighborhood and the minimum number of samples in a neighborhood for a point to be considered a point basic.

### CODE REVIEW

File: scripts/agnes.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    # Verifying if the number of missing values in the dataset is 0. (no missing value)
+    assert_equal(dataset.isna().sum().sum(), 0,
+                 err_msg=f'There is {dataset.isna().sum().sum()} NaN values in dataset, please preprocess them before '
+                         f'trying to fit AgglomerativeClustering.')
+
+    # Calculating pairwise distances between points in the dataset
+    dist_matrix = squareform(pdist(dataset, metric=dist_metric))
+
+    # Fitting the model
+    model = AgglomerativeClustering(n_clusters=n_clusters, linkage=linkage_method, affinity='precomputed')
```
Comment (HerySon) [inline scripts/agnes.py:41]: attention, le paramètre `affinity` est déprécié dans la version actuelle de scikit-learn, à la place il faut utliser `metric`. Dans ce cas il est plus simple de passer directement le dataframe des données à traiter en argument de la méthode `.fit` (pas besoin forcément de calculer `dist_matrix` à la ligne 38

File: scripts/agnes.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    model.fit(dist_matrix)
+
+    return model
```
Comment (HerySon) [inline scripts/agnes.py:44]: Il aurait été intressant d'appliquer au dataframe d'entrée les clusters attribués à chaque observation

File: scripts/dbscan.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    dbscan = DBSCAN(eps=eps, min_samples=min_samples)
+    labels = dbscan.fit_predict(dataset)
+    
+    # Calculate silhouette score, ignoring outliers (-1)
+    if -1 in labels:
+        silhouette = silhouette_score(dataset[labels!=-1], labels[labels!=-1])
+    else:
+        silhouette = silhouette_score(dataset, labels)
+    
+    return labels, silhouette
```
Comment (HerySon) [inline scripts/dbscan.py:31]: Il peut être utile aussi de renvoyer le modele entrainé, pour accéder si besoin à ses attributs

File: scripts/gmm.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from sklearn.mixture import GaussianMixture
+import matplotlib.pyplot as plt
+
+def gmm_clustering(dataset, n_components_range=(2,10)):
+    """
+    This function will fit a Gaussian Mixture Model. You pass the dataset as entry and you get the fitted model at the end.
```
Comment (HerySon) [inline scripts/gmm.py:6]: Scinder cette ligne qui est trop longue

File: scripts/gmm.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    dataset = dataset.dropna()
+
+    # Verifying if there are at least 2 rows after removing missing values
+    assert dataset.shape[0] >= 2, 'The dataset must contain at least 2 rows after removing missing values.'
+
+    # Verifying if all columns are numeric
+    assert all(dataset.dtypes.apply(lambda x: np.issubdtype(x, np.number))), 'All columns must be numeric.'
+
+    # Scaling the dataset
+    scaler = StandardScaler()
```
Comment (HerySon) [inline scripts/gmm.py:27]: proposer un argument optionnel pour choisir ou non un scaling préalable (inutile par exemple si les données ont déja été scalées)

File: scripts/gmm.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        bic.append(gmm.bic(dataset_scaled))
+
+    # Plotting the BIC as a function of the number of components
+    plt.plot(range(n_components_range[0], n_components_range[1]+1), bic, marker='o')
+    plt.xlabel("Number of components")
+    plt.ylabel("Bayesian Information Criterion")
+    plt.title("BIC as a function of number of components")
+
+    # Choosing the best number of components
+    best_n_components = np.argmin(bic) + n_components_range[0]
```
Comment (HerySon) [inline scripts/gmm.py:43]: :+1:

**Review by HerySon — CHANGES_REQUESTED** [review]

**Comment by HerySon** [comment]
@thomasleydet 

Bon travail ! il ne reste que quelques modifications très mineures à faires

## todo
- [ ] modifications mineures

**Review by HerySon — APPROVED** [review]
